lib/bolt/config/transport/base.rb

```diff
@@ -11,39 +11,42 @@ class Base
 
         def initialize(data = {}, boltdir = nil)
           assert_hash_or_config(data)
-          @input   = reference?(data) ? data : filter(data)
-          @config  = Bolt::Util.deep_merge(defaults, @input)
-          @boltdir = boltdir
-          validate
+          @input    = data
+          @resolved = !Bolt::Util.references?(input)
+          @config   = Bolt::Util.deep_merge(defaults, filter(input))
+          @boltdir  = boltdir
+
+          validate if resolved?
         end
 
+        # Accessor methods
+        # These are mostly all wrappers for same-named Hash methods, but they all
+        # require that the config options be fully-resolved before accessing them
         def [](key)
+          assert_resolved
           @config[key]
         end
 
         def to_h
+          assert_resolved
           @config
         end
 
         def fetch(*args)
+          assert_resolved
           @config.fetch(*args)
         end
 
         def include?(args)
+          assert_resolved
           @config.include?(args)
         end
 
         def dig(*keys)
+          assert_resolved
```

Comment:
Perhaps we could have a `resolved_config` method that asserts the data has been resolved+validated and then return `@config`. So this method, for instance, would just be:

```rb
def dig(*keys)
  resolved_config.dig(*keys)
end
```

We can also then be clear (by looking for `@config`) that we never use the unresolved config anywhere.